proof/refine/ARM/Ipc_R.thy

```diff
@@ -4109,101 +4536,298 @@ lemma replyPush_sch_act_wf:
+                                         \<longrightarrow>  obj_at' (\<lambda>ko. replyNext ko = Some (Head (fromJust scPtrOptDonated))) rp s)"
+                    in hoare_strengthen_post[rotated], clarsimp)
+       apply (wp hoare_vcg_all_lift hoare_vcg_imp_lift')
+      apply (simp (no_asm_use) add: getReplyTCB_def split del: if_split
+             | wp hoare_vcg_all_lift hoare_vcg_disj_lift hoare_vcg_imp_lift'
+                  hoare_vcg_if_lift setReply_obj_at' setReplyTCB_obj_at'
+             | rule threadGet_wp)+
+  apply clarsimp
+  apply (frule pred_tcb_at')
   apply (rule context_conjI; clarsimp simp: sym_refs_asrt_def)
+   apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_tcb projectKO_sc projectKO_reply)
    apply (erule if_live_then_nonz_capE')
-   apply (drule_tac ko=obj in sym_refs_ko_atD'[rotated])
+   apply (drule_tac ko=obja in sym_refs_ko_atD'[rotated])
     apply (fastforce simp: obj_at'_def projectKO_eq projectKO_tcb)
    apply (fastforce simp: live_sc'_def idle_tcb'_def valid_idle'_def
                           refs_of_rev' pred_tcb_at'_def ko_wp_at'_def obj_at'_def)
-  apply (fastforce elim: if_live_then_nonz_capE' dest: sym_refs_ko_atD'[rotated]
-                   simp: ko_wp_at'_def obj_at'_def projectKO_eq refs_of_rev' live_reply'_def)
+  apply (intro conjI allI)
+   apply (clarsimp simp: valid_tcb_state'_def)
+  apply (subgoal_tac "\<forall>rp scPtr. obj_at' (\<lambda>ko. scReply ko = Some rp) scPtr s \<longrightarrow>
+                                 obj_at' (\<lambda>ko. replyNext ko = Some (Head scPtr)) rp s")
+   apply (clarsimp simp: obj_at'_def)
+   apply (clarsimp)
+  apply (rule ReplySchedContext_obj_at'_ref)
+   apply (fastforce simp: sym_refs_def dest: obj_at'_SCReply_ref)
+  apply (frule_tac p=scPtr in  obj_at_ko_at', clarsimp)
+  apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+  apply (clarsimp simp: valid_sched_context'_def)
+  done
+
+lemma bindScReply_valid_idle':
+  "\<lbrace>valid_idle' and K (scPtr \<noteq> idle_sc_ptr)\<rbrace>
+   bindScReply scPtr replyPtr
+   \<lbrace>\<lambda>_. valid_idle'\<rbrace>"
+  unfolding bindScReply_def
+  by (wpsimp wp: hoare_vcg_imp_lift' hoare_vcg_all_lift setReply_obj_at')
+
+lemma bindScReply_obj_at'_scTCB[wp]:
+  "bindScReply scPtr replyPtr
+   \<lbrace>\<lambda>s. P (obj_at' (\<lambda>ko. P' (scTCB ko)) scPtr s)\<rbrace>"
+  unfolding bindScReply_def
+  apply (wpsimp wp: hoare_drop_imp setSchedContext_obj_at')
+  apply (auto simp: obj_at'_real_def ko_wp_at'_def)
+  done
+
+crunches setReplyTCB
+  for valid_idle'[wp]: valid_idle'
+
+(* This is a true weakest precondition, or at least very close *)
+lemma updateReply_valid_pspace'[wp]:
+  "\<lbrace>valid_pspace' and (\<lambda>s. \<forall>ko. ko_at' ko replyPtr s \<longrightarrow> valid_reply' (f ko) s)\<rbrace>
+   updateReply replyPtr f
+   \<lbrace>\<lambda>_. valid_pspace'\<rbrace>"
```

Comment:
Unfortunately, this will have a much more complicated precondition after I add `valid_replies'` to `invs'`. That's a problem for a later commit though.